File: go/cli/mcap/cmd/info.go
package cmd

import (
	"bufio"
	"bytes"
	"context"
	"fmt"
	"io"
	"math"
	"os"
	"sort"
	"strings"
	"time"

	"github.com/foxglove/mcap/go/cli/mcap/utils"
	"github.com/foxglove/mcap/go/mcap"
	"github.com/olekukonko/tablewriter"
	"github.com/spf13/cobra"
)

var (
	LongAgo = time.Now().Add(-20 * 365 * 24 * time.Hour)
)

func decimalTime(t time.Time) string {
	seconds := t.Unix()
	nanoseconds := t.Nanosecond()
	return fmt.Sprintf("%d.%09d", seconds, nanoseconds)
}

func humanBytes(numBytes uint64) string {
	prefixes := []string{"B", "KiB", "MiB", "GiB"}

	for index, p := range prefixes {
		displayedValue := float64(numBytes) / (math.Pow(1024, float64(index)))
		if displayedValue <= 1024 {
			return fmt.Sprintf("%.2f %s", displayedValue, p)
		}
	}
	lastIndex := len(prefixes) - 1
	displayedValue := float64(numBytes) / (math.Pow(1024, float64(lastIndex)))
	return fmt.Sprintf("%.2f %s", displayedValue, prefixes[lastIndex])
}

func getDurationNs(start uint64, end uint64) float64 {
	// subtracts start from end, returning the result as a 64-bit float.
	// float64 can represent the entire range of possible durations (-2**64, 2**64),
	// albeit with some loss of precision.
	diff := end - start
	signMultiplier := 1.0
	if start > end {
		diff = start - end
		signMultiplier = -1.0
	}
	return float64(diff) * signMultiplier
}

func addRow(rows [][]string, field string, value string, args ...any) [][]string {
	return append(rows, []string{field, fmt.Sprintf(value, args...)})
}

type chunkEvent struct {
	time             uint64
	isStart          bool
	uncompressedSize uint64
}

// countChunkOverlaps uses a sweep-line algorithm to find the maximum number
// of simultaneously active chunks and their total uncompressed size in O(n log n) time.
func countChunkOverlaps(chunks []*mcap.ChunkIndex) (
	hasOverlaps bool,
	maxActiveChunks int,
	maxTotalUncompressedSize uint64,
) {
	if len(chunks) < 2 {
		if len(chunks) == 1 {
			return false, 1, chunks[0].UncompressedSize
		}
		return false, 0, 0
	}

	// Create start and end events for each chunk
	events := make([]chunkEvent, 0, len(chunks)*2)
	for _, chunk := range chunks {
		events = append(events,
			chunkEvent{time: chunk.MessageStartTime, isStart: true, uncompressedSize: chunk.UncompressedSize},
			chunkEvent{time: chunk.MessageEndTime, isStart: false, uncompressedSize: chunk.UncompressedSize},
		)
	}

	// Sort events by time, with starts before ends at the same time
	sort.Slice(events, func(i, j int) bool {
		if events[i].time == events[j].time {
			// If times are equal, process starts before ends
			return events[i].isStart && !events[j].isStart
		}
		return events[i].time < events[j].time
	})

	// Sweep through events tracking max active chunks and total size
	activeChunks := 0
	maxActiveChunks = 0
	currentTotalSize := uint64(0)
	maxTotalUncompressedSize = 0

	for _, event := range events {
		if event.isStart {
			activeChunks++
			currentTotalSize += event.uncompressedSize
			if activeChunks > maxActiveChunks ||
				(activeChunks == maxActiveChunks && currentTotalSize > maxTotalUncompressedSize) {
				maxActiveChunks = activeChunks
				maxTotalUncompressedSize = currentTotalSize
			}
		} else {
			activeChunks--
			currentTotalSize -= event.uncompressedSize
		}
	}

	return maxActiveChunks > 1, maxActiveChunks, maxTotalUncompressedSize
}

func printInfo(w io.Writer, info *mcap.Info) error {
	buf := &bytes.Buffer{}

	header := [][]string{
		{"library:", info.Header.Library},
		{"profile:", info.Header.Profile},
	}
	var start, end uint64
	durationInSeconds := float64(0)
	if info.Statistics != nil {
		header = addRow(header, "messages:", "%d", info.Statistics.MessageCount)
		start = info.Statistics.MessageStartTime
		end = info.Statistics.MessageEndTime
		durationNs := getDurationNs(start, end)
		durationInSeconds = durationNs / 1e9
		starttime := time.Unix(int64(start/1e9), int64(start%1e9))
		endtime := time.Unix(int64(end/1e9), int64(end%1e9))
		if math.Abs(durationNs) > math.MaxInt64 {
			// time.Duration is an int64 nanosecond count under the hood, but end and start can
			// be further apart than that.
			header = addRow(header, "duration:", "%.3fs", durationInSeconds)
		} else {
			header = addRow(header, "duration:", "%s", endtime.Sub(starttime))
		}
		if starttime.After(LongAgo) {
			header = addRow(header, "start:", "%s (%s)", starttime.Format(time.RFC3339Nano), decimalTime(starttime))
			header = addRow(header, "end:", "%s (%s)", endtime.Format(time.RFC3339Nano), decimalTime(endtime))
		} else {
			header = addRow(header, "start:", "%s", decimalTime(starttime))
			header = addRow(header, "end:", "%s", decimalTime(endtime))
		}
	}
	if err := printSummaryRows(buf, header); err != nil {
		return err
	}
	if len(info.ChunkIndexes) > 0 {
		compressionFormatStats := make(map[mcap.CompressionFormat]struct {
			count            int
			compressedSize   uint64
			uncompressedSize uint64
		})

		hasOverlaps, maxActiveChunks, maxTotalUncompressedSize := countChunkOverlaps(info.ChunkIndexes)

		var largestChunkCompressedSize uint64
		var largestChunkUncompressedSize uint64

		for _, ci := range info.ChunkIndexes {
			stats := compressionFormatStats[ci.Compression]
			stats.count++
			stats.compressedSize += ci.CompressedSize
			stats.uncompressedSize += ci.UncompressedSize
			compressionFormatStats[ci.Compression] = stats

			if ci.CompressedSize > largestChunkCompressedSize {
				largestChunkCompressedSize = ci.CompressedSize
			}
			if ci.UncompressedSize > largestChunkUncompressedSize {
				largestChunkUncompressedSize = ci.UncompressedSize
			}
		}
		fmt.Fprintf(buf, "compression:\n")
		chunkCount := len(info.ChunkIndexes)
		for k, v := range compressionFormatStats {
			compressionRatio := 100 * (1 - float64(v.compressedSize)/float64(v.uncompressedSize))
			fmt.Fprintf(buf, "\t%s: [%d/%d chunks] ", k, v.count, chunkCount)
			fmt.Fprintf(buf, "[%s/%s (%.2f%%)] ",
				humanBytes(v.uncompressedSize), humanBytes(v.compressedSize), compressionRatio)
			if durationInSeconds > 0 {
				fmt.Fprintf(buf, "[%s/sec] ", humanBytes(uint64(float64(v.compressedSize)/durationInSeconds)))
			}
			fmt.Fprintf(buf, "\n")
		}
		fmt.Fprintf(buf, "chunks:\n")
		fmt.Fprintf(buf, "\tmax uncompressed size: %s\n", humanBytes(largestChunkUncompressedSize))
		fmt.Fprintf(buf, "\tmax compressed size: %s\n", humanBytes(largestChunkCompressedSize))
		if hasOverlaps {
			fmt.Fprintf(buf, "\toverlaps: [max concurrent: %d, decompressed: %s]\n",
				maxActiveChunks, humanBytes(maxTotalUncompressedSize))
		} else {
			fmt.Fprintf(buf, "\toverlaps: no\n")
		}
	}
	fmt.Fprintf(buf, "channels:\n")
	chanIDs := []uint16{}
	for chanID := range info.Channels {
		chanIDs = append(chanIDs, chanID)
	}
	sort.Slice(chanIDs, func(i, j int) bool {
		return chanIDs[i] < chanIDs[j]
	})
	rows := [][]string{}
	maxCountWidth := 0
	if info.Statistics != nil {
		for _, v := range info.Statistics.ChannelMessageCounts {
			count := fmt.Sprintf("%d", v)
			if len(count) > maxCountWidth {
				maxCountWidth = len(count)
			}
		}
	}

	maxChanIDWidth := 0
	if len(chanIDs) > 0 {
		maxChanIDWidth = digits(uint64(chanIDs[len(chanIDs)-1])) + 3
	}
	for _, chanID := range chanIDs {
		channel := info.Channels[chanID]
		schema := info.Schemas[channel.SchemaID]
		width := digits(uint64(chanID)) + 2
		padding := strings.Repeat(" ", maxChanIDWidth-width)
		row := []string{
			fmt.Sprintf("\t(%d)%s%s", channel.ID, padding, channel.Topic),
		}
		if info.Statistics != nil {
			channelMessageCount := info.Statistics.ChannelMessageCounts[chanID]
			if channelMessageCount > 1 {
				// Estimate frequency from statistics.
				// This assumes the underlying channel is logged at a quasi-constant rate with
				// random jitter.  Since we don't know where the MCAP start and end land in the
				// message's period, we calculate a lower bound (assuming the MCAP ends just
				// before the next message is logged) and an upper bound (assuming the MCAP ends
				// immediately after the last message was logged).
				// NOTE: We could make a better estimate by seeking to and inspecting message
				// indexes. However, these aren't present in all MCAPs, and seeking to each one can
				// be prohibitively slow, especially when accessing a remote file.
				seconds := float64(end-start) / 1e9
				maxHz := float64(channelMessageCount) / seconds
				minHz := float64(channelMessageCount-1) / seconds
				precision := int(max(0, math.Ceil(-math.Log10(maxHz-minHz))))
				row = append(
					row,
					fmt.Sprintf("%*d msgs (%.*f..%.*fHz)",
						maxCountWidth,
						channelMessageCount,
						precision,
						minHz,
						precision,
						maxHz),
				)
			} else {
				row = append(row, fmt.Sprintf("%*d msgs", maxCountWidth, channelMessageCount))
			}
		}
		switch {
		case schema != nil:
			row = append(row, fmt.Sprintf(" : %s [%s]", schema.Name, schema.Encoding))
		case channel.SchemaID != 0:
			row = append(row, fmt.Sprintf(" : <missing schema %d>", channel.SchemaID))
		default:
			row = append(row, " : <no schema>")
		}
		rows = append(rows, row)
	}
	if err := printSummaryRows(buf, rows); err != nil {
		return err
	}
	if info.Statistics != nil {
		fmt.Fprintf(buf, "channels: %d\n", info.Statistics.ChannelCount)
		fmt.Fprintf(buf, "attachments: %d\n", info.Statistics.AttachmentCount)
		fmt.Fprintf(buf, "metadata: %d\n", info.Statistics.MetadataCount)
	} else {
		fmt.Fprintf(buf, "channels: unknown\n")
		fmt.Fprintf(buf, "attachments: unknown\n")
		fmt.Fprintf(buf, "metadata: unknown\n")
	}
	_, err := buf.WriteTo(w)
	return err
}

// Similar to utils.FormatTable, but optimized for 'expanded' display of nested data.
func printSummaryRows(w io.Writer, rows [][]string) error {
	buf := &bytes.Buffer{}
	tw := tablewriter.NewWriter(buf)
	tw.SetBorder(false)
	tw.SetAutoWrapText(false)
	tw.SetAlignment(tablewriter.ALIGN_LEFT)
	tw.SetHeaderAlignment(tablewriter.ALIGN_LEFT)
	tw.SetColumnSeparator("")
	tw.AppendBulk(rows)
	tw.Render()
	// This tablewriter puts a leading space on the lines for some reason, so
	// remove it.
	scanner := bufio.NewScanner(buf)
	for scanner.Scan() {
		fmt.Fprintln(w, strings.TrimLeft(scanner.Text(), " "))
	}
	return scanner.Err()
}

var infoCmd = &cobra.Command{
	Use:   "info",
	Short: "Report statistics about an MCAP file",
	Run: func(_ *cobra.Command, args []string) {
		ctx := context.Background()
		if len(args) != 1 {
			die("Unexpected number of args")
		}
		// check if it's a remote file
		filename := args[0]
		err := utils.WithReader(ctx, filename, func(_ bool, rs io.ReadSeeker) error {
			reader, err := mcap.NewReader(rs)
			if err != nil {
				return fmt.Errorf("failed to get reader: %w", err)
			}
			defer reader.Close()
			info, err := reader.Info()
			if err != nil {
				return fmt.Errorf("failed to get info: %w", err)
			}
			err = printInfo(os.Stdout, info)
			if err != nil {
				return fmt.Errorf("failed to print info: %w", err)
			}
			return nil
		})
		if err != nil {
			die("Failed to read file %s: %v", filename, err)
		}
	},
}

func init() {
	rootCmd.AddCommand(infoCmd)
}

File: rust/mcap_cli/src/commands/info.rs
use std::collections::BTreeMap;
use std::fmt::Write as _;
use std::time::Duration;

use anyhow::Result;

use crate::cli::InfoCommand;
use crate::commands::common;
use crate::context::CommandContext;

pub fn run(_ctx: &CommandContext, args: InfoCommand) -> Result<()> {
    let mcap = common::map_file(&args.file)?;
    let parsed = common::parse_mcap(&mcap)?;
    print!("{}", render_info(&parsed));
    Ok(())
}

fn render_info(parsed: &common::ParsedMcap) -> String {
    let mut out = String::new();

    if let Some(header) = &parsed.header {
        let _ = writeln!(&mut out, "library: {}", header.library);
        let _ = writeln!(&mut out, "profile: {}", header.profile);
    } else {
        let _ = writeln!(&mut out, "library: unknown");
        let _ = writeln!(&mut out, "profile: unknown");
    }

    let mut duration_seconds = 0.0f64;
    if let Some(stats) = &parsed.statistics {
        let _ = writeln!(&mut out, "messages: {}", stats.message_count);
        let (duration_ns, signed_duration) =
            format_duration(stats.message_start_time, stats.message_end_time);
        duration_seconds = duration_ns / 1e9;
        let _ = writeln!(&mut out, "duration: {signed_duration}");
        let _ = writeln!(
            &mut out,
            "start: {}",
            common::formatted_time(stats.message_start_time)
        );
        let _ = writeln!(
            &mut out,
            "end: {}",
            common::formatted_time(stats.message_end_time)
        );
    }

    if !parsed.chunk_indexes.is_empty() {
        let mut by_compression: BTreeMap<&str, CompressionStats> = BTreeMap::new();
        let mut largest_compressed = 0u64;
        let mut largest_uncompressed = 0u64;

        for chunk in &parsed.chunk_indexes {
            let compression = chunk.compression.as_str();
            let stats = by_compression.entry(compression).or_default();
            stats.count += 1;
            stats.compressed_size += chunk.compressed_size;
            stats.uncompressed_size += chunk.uncompressed_size;

            largest_compressed = largest_compressed.max(chunk.compressed_size);
            largest_uncompressed = largest_uncompressed.max(chunk.uncompressed_size);
        }

        let (has_overlaps, max_active_chunks, max_total_uncompressed_size) =
            count_chunk_overlaps(&parsed.chunk_indexes);

        let _ = writeln!(&mut out, "compression:");
        let chunk_count = parsed.chunk_indexes.len();
        for (compression, stats) in by_compression {
            let ratio = if stats.uncompressed_size == 0 {
                0.0
            } else {
                100.0 * (1.0 - (stats.compressed_size as f64 / stats.uncompressed_size as f64))
            };

            let _ = write!(
                &mut out,
                "\t{compression}: [{}/{} chunks] [{}/{} ({ratio:.2}%)]",
                stats.count,
                chunk_count,
                human_bytes(stats.uncompressed_size),
                human_bytes(stats.compressed_size),
            );
            if duration_seconds > 0.0 {
                let throughput = (stats.compressed_size as f64 / duration_seconds).max(0.0);
                let _ = write!(&mut out, " [{}/sec]", human_bytes(throughput as u64));
            }
            let _ = writeln!(&mut out);
        }

        let _ = writeln!(&mut out, "chunks:");
        let _ = writeln!(
            &mut out,
            "\tmax uncompressed size: {}",
            human_bytes(largest_uncompressed)
        );
        let _ = writeln!(
            &mut out,
            "\tmax compressed size: {}",
            human_bytes(largest_compressed)
        );
        if has_overlaps {
            let _ = writeln!(
                &mut out,
                "\toverlaps: [max concurrent: {max_active_chunks}, decompressed: {}]",
                human_bytes(max_total_uncompressed_size)
            );
        } else {
            let _ = writeln!(&mut out, "\toverlaps: no");
        }
    }

    let _ = writeln!(&mut out, "channels:");
    let rows = render_channel_summary_rows(parsed, duration_seconds);
    out.push_str(&common::format_table(&rows));

    if let Some(stats) = &parsed.statistics {
        let _ = writeln!(&mut out, "channels: {}", stats.channel_count);
        let _ = writeln!(&mut out, "attachments: {}", stats.attachment_count);
        let _ = writeln!(&mut out, "metadata: {}", stats.metadata_count);
    } else {
        let _ = writeln!(&mut out, "channels: unknown");
        let _ = writeln!(&mut out, "attachments: unknown");
        let _ = writeln!(&mut out, "metadata: unknown");
    }

    out
}

fn render_channel_summary_rows(
    parsed: &common::ParsedMcap,
    duration_seconds: f64,
) -> Vec<Vec<String>> {
    let mut rows = Vec::new();
    for channel in parsed.channels.values() {
        let msg_col = if let Some(stats) = &parsed.statistics {
            let count = stats
                .channel_message_counts
                .get(&channel.id)
                .copied()
                .unwrap_or_default();
            if count > 1 && duration_seconds > 0.0 {
                let max_hz = count as f64 / duration_seconds;
                let min_hz = (count - 1) as f64 / duration_seconds;
                let delta = (max_hz - min_hz).max(f64::EPSILON);
                let precision = ((-delta.log10()).ceil() as i32).max(0) as usize;
                if precision > 2 {
                    format!("{count} msgs ({max_hz:.2}Hz)")
                } else {
                    format!("{count} msgs ({min_hz:.precision$}..{max_hz:.precision$}Hz)")
                }
            } else {
                format!("{count} msgs")
            }
        } else {
            String::new()
        };

        let schema_col = match channel.schema_id {
            0 => " : <no schema>".to_string(),
            id => match parsed.schemas.get(&id) {
                Some(schema) => {
                    format!(" : {} [{}]", schema.header.name, schema.header.encoding)
                }
                None => format!(" : <missing schema {id}>"),
            },
        };

        rows.push(vec![
            format!("\t({}) {}", channel.id, channel.topic),
            msg_col,
            schema_col,
        ]);
    }
    rows
}

fn format_duration(start: u64, end: u64) -> (f64, String) {
    let (diff, sign) = if end >= start {
        (end - start, "")
    } else {
        (start - end, "-")
    };
    let duration = Duration::from_nanos(diff);
    (diff as f64, format!("{sign}{duration:?}"))
}

fn human_bytes(num_bytes: u64) -> String {
    let prefixes = ["B", "KiB", "MiB", "GiB"];
    for (index, prefix) in prefixes.iter().enumerate() {
        let displayed = num_bytes as f64 / 1024f64.powi(index as i32);
        if displayed <= 1024.0 {
            return format!("{displayed:.2} {prefix}");
        }
    }

    let last = prefixes.len() - 1;
    let displayed = num_bytes as f64 / 1024f64.powi(last as i32);
    format!("{displayed:.2} {}", prefixes[last])
}

#[derive(Default)]
struct CompressionStats {
    count: usize,
    compressed_size: u64,
    uncompressed_size: u64,
}

#[derive(Debug, Clone, Copy)]
struct ChunkEvent {
    time: u64,
    is_start: bool,
    uncompressed_size: u64,
}

fn count_chunk_overlaps(chunks: &[mcap::records::ChunkIndex]) -> (bool, usize, u64) {
    if chunks.len() < 2 {
        if let Some(chunk) = chunks.first() {
            return (false, 1, chunk.uncompressed_size);
        }
        return (false, 0, 0);
    }

    let mut events = Vec::with_capacity(chunks.len() * 2);
    for chunk in chunks {
        events.push(ChunkEvent {
            time: chunk.message_start_time,
            is_start: true,
            uncompressed_size: chunk.uncompressed_size,
        });
        events.push(ChunkEvent {
            time: chunk.message_end_time,
            is_start: false,
            uncompressed_size: chunk.uncompressed_size,
        });
    }

    events.sort_by(|a, b| {
        a.time
            .cmp(&b.time)
            .then_with(|| b.is_start.cmp(&a.is_start))
    });

    let mut active = 0usize;
    let mut max_active = 0usize;
    let mut current_size = 0u64;
    let mut max_size = 0u64;

    for event in events {
        if event.is_start {
            active += 1;
            current_size += event.uncompressed_size;
            if active > max_active || (active == max_active && current_size > max_size) {
                max_active = active;
                max_size = current_size;
            }
        } else {
            active = active.saturating_sub(1);
            current_size = current_size.saturating_sub(event.uncompressed_size);
        }
    }

    (max_active > 1, max_active, max_size)
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use super::{count_chunk_overlaps, human_bytes, render_info};
    use crate::commands::common;
    use crate::commands::common::{ParsedMcap, ParsedSchema};
    use mcap::records::{self, ChunkIndex, Header, Statistics};

    #[test]
    fn decimal_time_formats_nanos() {
        assert_eq!(common::decimal_time(1_234_567_890), "1.234567890");
    }

    #[test]
    fn formatted_time_includes_rfc3339_and_decimal() {
        assert_eq!(
            common::formatted_time(1_000_000_000),
            "1970-01-01T00:00:01.000000000Z (1.000000000)"
        );
    }

    #[test]
    fn human_bytes_scales_units() {
        assert_eq!(human_bytes(2), "2.00 B");
        assert_eq!(human_bytes(2 * 1024), "2.00 KiB");
    }

    #[test]
    fn overlaps_counts_concurrent_chunks() {
        let chunks = vec![
            ChunkIndex {
                message_start_time: 10,
                message_end_time: 20,
                chunk_start_offset: 0,
                chunk_length: 0,
                message_index_offsets: BTreeMap::new(),
                message_index_length: 0,
                compression: "zstd".to_string(),
                compressed_size: 10,
                uncompressed_size: 100,
            },
            ChunkIndex {
                message_start_time: 15,
                message_end_time: 25,
                chunk_start_offset: 0,
                chunk_length: 0,
                message_index_offsets: BTreeMap::new(),
                message_index_length: 0,
                compression: "zstd".to_string(),
                compressed_size: 15,
                uncompressed_size: 80,
            },
        ];
        let (has_overlaps, max_active, max_size) = count_chunk_overlaps(&chunks);
        assert!(has_overlaps);
        assert_eq!(max_active, 2);
        assert_eq!(max_size, 180);
    }

    #[test]
    fn info_render_includes_core_sections() {
        let mut parsed = ParsedMcap {
            header: Some(Header {
                profile: "demo".to_string(),
                library: "mcap-rust".to_string(),
            }),
            statistics: Some(Statistics {
                message_count: 2,
                channel_count: 1,
                attachment_count: 0,
                metadata_count: 0,
                chunk_count: 1,
                message_start_time: 1_000_000_000,
                message_end_time: 2_500_000_000,
                ..Statistics::default()
            }),
            ..ParsedMcap::default()
        };
        parsed.schemas.insert(
            1,
            ParsedSchema {
                header: records::SchemaHeader {
                    id: 1,
                    name: "demo_schema".to_string(),
                    encoding: "jsonschema".to_string(),
                },
                data: br#"{"type":"object"}"#.to_vec(),
            },
        );
        parsed.channels.insert(
            7,
            records::Channel {
                id: 7,
                schema_id: 1,
                topic: "/demo".to_string(),
                message_encoding: "json".to_string(),
                metadata: BTreeMap::new(),
            },
        );

        let rendered = render_info(&parsed);
        assert!(rendered.contains("library: mcap-rust"));
        assert!(rendered.contains("profile: demo"));
        assert!(rendered.contains("messages: 2"));
        assert!(rendered.contains("channels:"));
        assert!(rendered.contains("/demo"));
    }
}
